content/articles/2017-11-01-lubridate-1.7.0.Rmd

```diff
@@ -0,0 +1,135 @@
+
+## Vectorized time-zone manipulation
+
+Often date-time data comes with heterogeneous time-zones. For example, you might
+have a character vector of date-times and a separate vector of time-zones. In
+such cases you can parse your date-time vector in one time zone (`UTC`) and then
+"enforce" the heterogeneous time-zones with `force_tzs`.
+
+
+```{r}
+x <- ymd_hms(c("2009-08-01 00:00:00", "2009-08-01 00:00:00"))
+tzs <- c("America/New_York", "Europe/Amsterdam")
+
+force_tzs(x, tzones = tzs)
+
+force_tzs(x, tzones = tzs, tzone_out = "America/New_York")
+```
+
+Note that first `force_tzs` call produced a vector of instants in UTC time-zone,
+for the second call we specified the desired time-zone of the output vector with
+the `tzone_out` argument. This is needed as R's date-time vectors cannot
+represent heterogeneous time-zones. For the same reason, the counterpart of
+`with_tz`, `with_tzs`, does not exist. Instead, `local_time` should cover the
+`with_tzs` use case in most situations.
+
+New `local_time` function retrieves day clock time of the input vector in the
+specified time zones. It returns a `difftime` object and it is vectorized over
+both date-time and time-zone arguments:
+
+```{r}
+x <- ymd_hms(c("2009-08-01 01:02:03", "2009-08-01 10:20:30"))
+local_time(x, units = "hours")
+
+x <- ymd_hms(c("2009-08-01 00:00:00", "2009-08-01 00:00:00"))
+tzs <- c("America/New_York", "Europe/Amsterdam")
+local_time(x, tzs, units = "hours")
+```
+
+## New `cutoff_2000` parameter in parsing functions
+
+Lower level lubridate parser functions (`parse_datetime2`,`fast_strptime`) now
+accept `cutoff_2000` parameter to determine when the parsing of `yy` format
+should output 19th or 20th century.
+
+```{r}
+x <- c("50-01-01", "70-01-01")
+parse_date_time2(x, "ymd")
+parse_date_time2(x, "ymd", cutoff_2000 = 30)
+```
+
+By default `cutoff_2000` is 68 to comply with R's `strptime` function.
+
+This feature was not propagated to higher level parsing function because base
+`strptime`, on which those occasionally rely, doesn't support this option.
+
+
+## Functions `wday` and `month` are now localized
+
+When `label=TRUE` `wday` and `month` extractors return localized strings:
+
+```{r}
+Sys.setlocale(locale = "zh_CN.utf8")
+wday(now(), label = TRUE)
+month(now(), label = TRUE)
+```
+
+## Several functions gained the `week_start` parameter
+
+Functions for which depend on the week-start conventions (`wday`, `wday<-`,
+`floor_date`, `ceiling_date` and `round_date`) now accept `week_start` argument
+which defaults to `getOption("lubridate.week.start", 7)`.
+
+```{r}
+x <- today()
+wday(x, label = TRUE, abbr = FALSE)
+wday(x)
+wday(x,  week_start = 2)
+```
+
+## Rounding to fraction of a second
+
+R's date-time format (POSIXct) is accurate up to microseconds, thus it makes
+sense to round to fractions of a second. Unfortunately R currently prints
+fractional seconds incorrectly which can lead to confusion:
```

Comment:
Maybe say "truncates" instead of "rounds" rather than incorrect?

Reply:
It's a printing issue. I don't know the exact nature of the bug but I don't think it's as simple as truncation. As this example ilustrates R prints ".02" instead of ".03", so it's really an incorrect printing of decimals. No time to dig it further but I have reported it [here](https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17350).

Reply:
`format.POSIXct` has:

> Specific to R is %OSn, which for output gives the seconds truncated to 0 <= n <= 6 decimal places (and if %OS is not followed by a digit, it uses the setting of getOption("digits.secs"), or if that is unset, n = 0).

Reply:
So it's technically not a bug, but documented behaviour 🙄 which means R-core is unlikely to fix it. I suspect they would accept a patch though, and I suspect it's probably not a lot of work.

Reply:
That doc just states that it truncates to digits.secs. In my example digits.secs is set to 6, so it should print .03 as .03 not as .02. It's really a bug.

Reply:
Well there's a bug that it doesn't respect `digits.secs`, but the truncation is by design. 

(This is all buggy behaviour but describing it as a bug to R-core when it's (semi) documented will trigger them)